### PR TITLE
store a copy of the state usage for gas schedule

### PR DESCRIPTION
**Pull request by msmouse**
### Description

This doesn't change during the epoch so safe to be read from the vm adaptor

### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3414)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/reconfiguration.move [hunk 1]
```diff
@@ -113,6 +114,7 @@ module aptos_framework::reconfiguration {
 
         // Call stake to compute the new validator set and distribute rewards.
         stake::on_new_epoch();
```
Comment (vgao1996) [inline aptos-move/framework/aptos-framework/sources/reconfiguration.move:116]: Unrelated to this but are we supposed to call `stake::on_new_epoch()` here? @movekevin

**Review by vgao1996 — APPROVED** [review]
LGTM

File: aptos-move/framework/aptos-framework/sources/reconfiguration.move (hunk 1 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/reconfiguration.move:116]: Yep this is correct. We need to recompute the validator set when there's a reconfig

File: aptos-move/framework/aptos-framework/sources/state_storage.move [hunk 2]
```diff
@@ -1,59 +1,82 @@
...
+    ///
+    /// Specifically, it copies the usage at the begining of the concluding
+    /// epoch for gas calculation of the entire next epoch. -- The data is one
+    /// epoch older than ideal, but the Vm doesn't need to worry about reloading
+    /// gas parameters after the first txn of an epoch.
+    struct GasParameter has key, store {
+        usage: Usage,
     }
 
+
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/state_storage.move:37]: nit: Unnecessary empty line

File: aptos-move/framework/aptos-framework/sources/state_storage.move [hunk 3]
```diff
@@ -1,59 +1,82 @@
...
         items: u64,
         bytes: u64,
     }
 
-    struct Usage_ {
-        items: u64,
-        bytes: u64,
+    /// This is updated at the begining of each opoch, reflecting the storage
+    /// usage after the last txn of the previous epoch is committed.
+    struct StateStorageUsage has key, store {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/state_storage.move:21]: If I understand correctly, you write to StateStorageUsage in the first block of the new epoch and to GasParameter in the last block of the previous epoch. This seems redundant. Why do we need to do this?

**Review by movekevin — CHANGES_REQUESTED** [review]

File: aptos-move/framework/aptos-framework/sources/state_storage.move (hunk 3 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/state_storage.move:21]: > you write to StateStorageUsage in the first block of the new epoch

the comment on `native fun get_state_storage_usage_only_at_epoch_beginning()` -- the only occasion every node is guaranteed to be able to read it

> and to GasParameter in the last block of the previous epoch

the comment on `struct GasParameter has key, store` -- the vm needs the value to not change except in the last txn in an epoch

**Review by movekevin — APPROVED** [review]

**Review by zekun000 — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/state_storage.move (hunk 3 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/state_storage.move:21]: Talked to Alden offline and this makes sense now 👍

**Pull request merged**